### ISSUE TITLE
error "Invalid path part "" on uploaded coverage

### ISSUE THREAD

**Issue by zanona** [codeclimate/test-reporter#375]
Hey, guys
I am having issues while uploading coverage locally.
After `cc-test-reporter after-build` the coverage is successfully uploaded but I do get an error under code climate mentioning the coverage is invalid.

![image](https://user-images.githubusercontent.com/32017/48486101-d2e7cb00-e81a-11e8-9d0c-8c6563b992bd.png)

error: (formatted for readability)

```json
{
  "message": "Invalid path part \"\"",
  "file_document": "{
    \"_id\"=>BSON::ObjectId('5bec2222e586504995006bd8'), 
    \"type\"=>\"test_file_reports\",
    \"blob_id\"=>\"8f2152d437a409b348c438b3d72059e8bbba1262\", 
    \"coverage\"=>\"[1,1]\",
    \"covered_percent\"=>100, 
    \"covered_strength\"=>1, 
    \"line_counts\"=>   {\"missed\"=>0, \"covered\"=>2, \"total\"=>2},
    \"path\"=>\"/mnt/c/Users/Marcus/Developer/project/lib/utils.ts\",
    \"test_report_id\"=>BSON::ObjectId('5bec22210fef2e304f004f73')
   }"
}
```

my test command looks like:

```bash
npx nyc ava-ts
npx nyc report --reporter lcov
cc-test-reporter after-build
```

any ideas?

**Comment by zanona** [comment]
Thanks for following up, Will.
So, my environment is as you mentioned indeed. I am running WSL as Ubuntu 16.04.5 LTS.

```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description: Ubuntu 16.04.5 LTS
Release: 16.04
Codename: xenial

$ bash --version
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law. 

$ node -v
v9.11.2
```
I hope this can be helpful. Please let me know I can provide you with any other info.

Thanks

**Comment by wfleming** [comment]
The default `--prefix` is the working directory where you're running the command from. Why that is different than the path Istanbul seemed to decided to use is odd, and I don't know why it would do that.

I'd suggest trying to understand why the istanbul report is using absolute paths at all, and seeing if you can configure it to output relative paths itself. That should fix the usage with Code Climate, and I is also likely to save you some headaches in the future, since the report with absolute paths is likely to run into similar issues if you ever want to use it with any other 3rd party tools or services.

That said, if you just want to move forward with getting the report to work on Code Climate to work, you can use the `--prefix` option to do that. That option is used to strip excessive pieces off of path names off of some coverage reports in cases like this, but the mismatch between your actual working path and the path that Istanbul seems to be seeing is the problem. If you used `--prefix /mnt/c/Users/Marcus/Developer/project/`, it should strip the paths to be correctly relative to the project root.

**Comment by zanona** [comment]
Thanks again for having a look at this, Will.
So, I tried to use the prefix option as you have suggested and it worked - thanks for that 👍 

After analysing further, I could spot the difference between `nyc` when analysing the cwd and `cc-test-reporter` it may have something to do with retrieving the real path by reading the link to the directory rather than the system pwd?

have a look:

### pwd
```
$ pwd
/home/zanona/Developer/project
```
### readlink
```
$ readlink -f .
/mnt/c/Users/Marcus/Developer/project
```
I believe that the second example routes the original path of the file instead of the alias that is created. I believe that perhaps, regardless of the configuration, `readlink` is something that would always work, whether `pwd` may have its drawbacks? 

As a workaround, I am now calling `cc-test-reporter after-build --prefix $(readlink -f .)`

What do you think? And, do you believe that perhaps using something like `readlink` to retrieve the current directory could perhaps work as a replacement for the actual way of doing it?

Thanks again.

**Comment by wfleming** [comment]
Ah. Symlinks. Good call.

I don't think we're going to want to change the determination of "default" `--prefix` on our end, at least not without more evidence that a lot more coverage tools are fully canonicalizing paths by following symlinks, as Istanbul seems to be doing. This is one of those situations where I fully expect some tools use the canonicalized path, and some coverage tools use the "apparent" (symlinked) path, so if we changed this to fix this case I fully expect it would break some other configurations, and I think either choice is a defensible one, so I'm inclined to not change the existing behavior in our test-reporter.

That said I would consider a change on our end for the test reporter to validate file paths aren't still absolute after prefix operations have occurred, which would at least have caught this problem earlier and could have provided a more logical error message.

I'm going to go ahead and close this since we discovered what was affecting your configuration and found a config that works for you, but feel free to still comment here if you have further questions or want to discuss more.
